server/odc-service/src/main/java/com/oceanbase/odc/service/iam/auth/JwtSecurityContextRepository.java

```diff
@@ -0,0 +1,174 @@
+            return context;
+        }
+
+        Date expiration = jwtService.getExpiresAt(token);
+        if (jwtService.isExpired(expiration)) {
+            log.info("Created SecurityContext {} because token is expired", context);
+            return context;
+        }
+
+        Map<String, Claim> claims = jwtService.getClaims(token);
+        long id = claims.get(JwtConstants.ID).asLong();
+
+        if (authenticationCache.getIfPresent(id) != null) {
+            Authentication authentication = authenticationCache.getIfPresent(id);
+            context.setAuthentication(authentication);
+        } else {
+            String username = claims.get(JwtConstants.PRINCIPAL).asString();
+            UserEntity userEntity = userRepository.findByAccountName(username).orElseThrow(() -> {
+                log.warn("Username not found: username {}", username);
+                return new UsernameNotFoundException(username);
+            });
+            User user = new User(userEntity);
+            user.setOrganizationId(claims.get(JwtConstants.ORGANIZATION_ID).asLong());
+
+            OrganizationType organizationType =
+                    JsonUtils.fromJson(claims.get(JwtConstants.ORGANIZATION_TYPE).asString(),
+                            OrganizationType.class);
+            user.setOrganizationType(organizationType);
+            UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken =
+                    new UsernamePasswordAuthenticationToken(user, null, null);
+            context.setAuthentication(usernamePasswordAuthenticationToken);
+            authenticationCache.put(user.getId(), usernamePasswordAuthenticationToken);
+        }
+
+        // jwt renewal policy, the default is to complete the renewal before the expiration time, to avoid
+        // frequent updates
+        User user = (User) context.getAuthentication().getPrincipal();
+        if (jwtService.isRenew(expiration)) {
+            HashMap<String, Object> hashMap = new HashMap<>();
+            hashMap.put(JwtConstants.ID, user.getId());
+            hashMap.put(JwtConstants.PRINCIPAL, user.getAccountName());
+            hashMap.put(JwtConstants.ORGANIZATION_ID, user.getOrganizationId());
+            hashMap.put(JwtConstants.ORGANIZATION_TYPE, JsonUtils.toJson(user.getOrganizationType()));
+            String renewToken = jwtService.sign(hashMap);
+            Cookie cookie = new Cookie(JwtConstants.ODC_JWT_TOKEN, renewToken);
+            cookie.setPath("/");
+            cookie.setMaxAge((int) timeoutSetting.getSeconds());
+            cookie.setHttpOnly(true);
+            response.addCookie(cookie);
+        }
+        return context;
+    }
+
+    @Override
+    public void saveContext(SecurityContext context, HttpServletRequest request, HttpServletResponse response) {
+        if (request.getRequestURI().contains(commonSecurityProperties.getLoginUri())
```

Comment:
assume use bastion login, the login uri will not match commonSecurityProperties.getLoginUri.
then saveContext will not works.

may ispired from HttpSessionSecurityContextRepository implmentation, judge from session status instead.
![image](https://github.com/oceanbase/odc/assets/54161870/6b7cdde4-729e-4962-8653-636a041d159e)

or, can we just do nothing while saveContext?